c/python/namedtuple.cc

```diff
@@ -27,50 +27,68 @@
-*  However, objects created with `PyStructSequence_New` are not standard
-*  Python namedtuples, but rather a simpler version of the latter.
-*/
-onamedtupletype::onamedtupletype(strpair tuple_info,
-                                 std::vector<strpair> fields_info) {
+ * Create a namedtuple using python function `collections.namedtuple()`.
+ *
+ * Note: we do not use PyStructSequence objects from Python C API, because
+ * they are not standard namedtuples and do not provide the same API (for
+ * example, there is no `_replace()` method).
+ * https://docs.python.org/3/c-api/tuple.html
+ */
+onamedtupletype::onamedtupletype(const std::string& cls_name,
+                                 const std::string& cls_doc,
+                                 const strvec& field_names,
+                                 const strvec& field_docs)
+{
   auto itemgetter = py::oobj::import("operator", "itemgetter");
   auto namedtuple = py::oobj::import("collections", "namedtuple");
   auto property   = py::oobj::import("builtins", "property");
 
   // Create a namedtuple type from the supplied fields
-  nfields = fields_info.size();
+  nfields = field_names.size();
   py::olist argnames(nfields);
   for (size_t i = 0; i < nfields; ++i) {
-    argnames.set(i, py::ostring(fields_info[i].first));
+    argnames.set(i, py::ostring(field_names[i]));
   }
 
   py::otuple args(2);
-  args.set(0, py::ostring(tuple_info.first));
+  args.set(0, py::ostring(cls_name));
   args.set(1, argnames);
 
   auto type = namedtuple.call(args);
   auto res = std::move(type).release();
 
   // Set namedtuple doc
-  py::ostring docstr = py::ostring(tuple_info.second);
-  PyObject_SetAttrString(res, "__doc__",
-                         docstr.to_borrowed_ref());
+  if (!cls_doc.empty()) {
+    py::ostring docstr = py::ostring(cls_doc);
+    PyObject_SetAttrString(res, "__doc__",
+                           docstr.to_borrowed_ref());
+  }
 
   // Set field docs
-  py::otuple args_prop(4);
-  py::otuple args_itemgetter(1);
-  args_prop.set(1, py::None());
-  args_prop.set(2, py::None());
-  for (size_t i = 0; i < nfields; ++i) {
-    args_itemgetter.set(0, py::oint(i));
-    args_prop.set(0, itemgetter.call(args_itemgetter));
-    args_prop.set(3, py::ostring(fields_info[i].second));
-    auto prop = property.call(args_prop);
-    PyObject_SetAttrString(res, fields_info[i].first,
-                           prop.to_borrowed_ref());
+  if (!field_docs.empty()) {
+    py::otuple args_prop(4);
+    py::otuple args_itemgetter(1);
+    args_prop.set(1, py::None());
+    args_prop.set(2, py::None());
+    for (size_t i = 0; i < field_docs.size(); ++i) {
+      args_itemgetter.set(0, py::oint(i));
+      args_prop.set(0, itemgetter.call(args_itemgetter));
+      args_prop.set(3, py::ostring(field_docs[i]));
+      auto prop = property.call(args_prop);
+      PyObject_SetAttrString(res, field_names[i].c_str(),
+                             prop.to_borrowed_ref());
+    }
```

Comment:
If we don't use `strpair` to pass field names and field docs, then number of names could be different than the number of docs. For example, there is nothing preventing `field_docs.size()` to be larger than `nfields`. Do you know what will happen in this case?

Reply:
Any particular reason why you prefer not to use `std::pair`?

Reply:
Yes, because I wanted to support a common use-case when there are no doc-strings. For example, consider converting a `Frame` into a list of namedtuples. In this case there are only the names of the fields. Of course, I could have created strpairs with empty docstrings, but this seems like a rather awkward transform for no good reason.

Reply:
Though on a second thought, we could have used a `struct { std::string name, doc; }` there and achieved the same result provided the struct had suitable set of constructors.

Reply:
I see, but don't you have to create an empty `field_docs` now? Due to this `xassert(field_docs.size() == nfields)`, it seems that you require `field_docs` to have the same size as `field_names`, i.e. one need to create an empty `std::vector` of doc strings.

Reply:
Ah, sorry missed the `!field_docs.empty()` condition. Gotcha.

Reply:
Yes, that's also a possibility...

Reply:
How about now?

Reply:
Like it much better now.